File: 2024/37/onsen.rb
#!/usr/bin/env ruby

# 2024 Day 19 Part 1 / Puzzle 2024/37

# quick 'n' dirty solution to get first star, will rewrite with Rust later but
# for now Ruby makes this easy.

class Puzzle
  def initialize(hdr, designs)
    @towels = hdr.split(', ').sort
    @designs = designs.split("\n")
    @stored_counts = { }
  end

  def num_matches(design)
    count = 0
    if @stored_counts.has_key?(design)
      return @stored_counts[design]
    end

    @towels.each { |t|
      if design === t
        # base case
        count += 1
        break
      end

      if design.start_with? t
        rest = design[(t.length)..]
        rem_count = num_matches(rest)
        if rem_count > 0
          count += rem_count
        end
      end
    }

    @stored_counts[design] = count
    return count
  end

  def split(stone, bleft)
    return stone.length unless bleft>0

    count = 0
    stone.each { |s|
      new_vals = []
      key = "#{bleft}-#{s}"

      if @stored_counts.has_key?(key)
        count += @stored_counts[key]
        next
      end

      if s == 0
        new_vals.push(1)
      elsif s.to_s.length % 2 == 0
        s1 = s.to_s
        new_vals = [ s1[0...(s1.length/2)].to_i, s1[(s1.length/2)..].to_i ]
      else
        new_vals.push(s * 2024);
      end

      @stored_counts[key] = self.split(new_vals, bleft - 1)
      count += @stored_counts[key]
    }

    return count
  end

  def designs
    @designs
  end

  def towels
    @towels
  end
end

hdr, designs = File.read(ARGV.shift || '../37/input').split("\n\n")

p = Puzzle.new(hdr, designs)

num_possible = p.designs.filter { |d| p.num_matches(d) > 0 }.count
puts "#{num_possible} are possible"

File: 2024/37/onsen.rs
use std::env;
use std::fs;
use std::collections::HashMap;

// Advent of Code: 2024, day 19, part 1

// Goal is to find number of ways to craft towel designs based on available
// towel subpart options
type DesignMap = HashMap<String, u64>;

fn count_options(design: &str, towels: &Vec<&str>, mut cache: &mut DesignMap) -> u64 {
    let mut count = 0u64;
    if cache.contains_key(design) {
        return cache[design];
    }

    for t in towels.iter() {
        if *t == design {
            count += 1;
            break;
        }

        if design.starts_with(t) {
            let rest = design.strip_prefix(t).unwrap();
            count += count_options(rest, &towels, &mut cache);
        }
    }

    cache.insert(String::from(design), count);
    return count;
}

fn split_at_empty_line(mut lines: Vec<String>) -> (Vec<String>, Vec<String>) {
    let idx = lines.iter().position(|el| el.is_empty()).unwrap();

    // +1 so we can pop off the empty line
    let designs = lines.split_off(idx+1);
    lines.pop();

    return (lines, designs);
}

fn main() {
    let default_filename: &'static str = "../37/input";
    let args: Vec<String> = env::args().collect();
    let in_file: String = match args.len() {
        // No args provided save argv[0]
        1 => String::from(default_filename),
        _ => args[1].clone(),
    };

    let lines = fs::read_to_string(&in_file)
        .expect("Should have been able to read the file")
        .lines()
        .map(String::from)
        .collect::<Vec<_>>();

    let (hdr, designs) = split_at_empty_line(lines);

    let mut towels = hdr[0].split(", ").collect::<Vec<_>>();
    towels.sort(); // This seems to be load-bearing!

    let mut cache = DesignMap::new();

    let num_options: usize = designs.iter()
        .map(|d| count_options(d, &towels, &mut cache))
        .filter(|count| *count > 0u64)
        .count();

    println!("{} total designs possible", num_options);
}
